orangecontrib/prototypes/widgets/owexplpred.py

```diff
@@ -0,0 +1,541 @@
+                # update progress bar
+                QMetaObject.invokeMethod(
+                    self, "set_progress_value", Qt.QueuedConnection, Q_ARG(int, progress))
+                return False
+
+            def callback_update(table):
+                QMetaObject.invokeMethod(
+                    self, "update_view", Qt.QueuedConnection, Q_ARG(Orange.data.Table, table))
+
+            def callback_prediction(class_value):
+                QMetaObject.invokeMethod(
+                    self, "update_model_prediction", Qt.QueuedConnection, Q_ARG(float, class_value))
+
+            self.was_canceled = False
+            explain_func = partial(
+                self.e.anytime_explain, self.to_explain[0], callback=callback, update_func=callback_update, update_prediction=callback_prediction)
+            task.future = self._executor.submit(explain_func)
+            task.watcher = FutureWatcher(task.future)
+            task.watcher.done.connect(self._task_finished)
+            self.cancel_button.setDisabled(False)
+            self.progressBarInit(processEvents=None)
+
+    @pyqtSlot(Orange.data.Table)
+    def update_view(self, table):
+        self.explanations = table
+        model = TableModel(table, parent=None)
+        header = self.dataview.horizontalHeader()
+        model.sort(
+            header.sortIndicatorSection(),
+            header.sortIndicatorOrder())
+        self.dataview.setModel(model)
+        self.commit_output()
+
+    @pyqtSlot(float)
+    def update_model_prediction(self, value):
+        self._print_prediction(value)
+
+    @pyqtSlot(int)
+    def set_progress_value(self, value):
+        self.progressBarSet(value, processEvents=False)
+
+    @pyqtSlot(concurrent.futures.Future)
+    def _task_finished(self, f):
+        """
+        Parameters:
+        ----------
+        f: conncurent.futures.Future
+            future instance holding the result of learner evaluation
+        """
+        assert self.thread() is QThread.currentThread()
+        assert self._task is not None
+        assert self._task.future is f
+        assert f.done()
+
+        self._task = None
+        self.progressBarFinished(processEvents=False)
+
+        if not self.was_canceled:
+            self.cancel_button.setDisabled(True)
+
+        try:
+            results = f.result()
+        except Exception as ex:
+            log = logging.getLogger()
+            log.exception(__name__, exc_info=True)
+            self.error("Exception occured during evaluation: {!r}".format(ex))
+
+            for key in self.results.keys():
+                self.results[key] = None
+        else:
+            self.explanations = results[1]
+            self.commit_output()
+            model = TableModel(self.explanations, parent=None)
+            header = self.dataview.horizontalHeader()
+            model.sort(
+            header.sortIndicatorSection(),
+            header.sortIndicatorOrder())
+            self.dataview.setModel(model)
```

Comment:
couldn't most of this `else` be replaced with `self.update_view(results[1])`

Reply:
Yes, all of it actually, thank you.